File: Entities.py
from GlobalVariables import blockSize, gravity, maxStackSize, itemEntitySize
from Worldgen import findBlock
import math, pygame, random

# basic entity, no ai or anything
class Entity():
    def __init__(self, x = 0, y = 0, z = 0, xv = 0, yv = 0, zv = 0):
        self.x = x
        self.y = y
        self.z = z

        self.xv = xv
        self.yv = yv
        self.zv = zv
        self.deleteSelf = False

        self.width = blockSize
        self.height = self.width

        self.rect = pygame.rect.Rect(0, 0, self.width, self.width)

# used for items that are on the ground, like after breaking something
class ItemEntity(Entity):
    def __init__(self, itemData, count, x, y, z, xv = 0, yv = 0, zv = 0):
        super().__init__(x, y, z, xv, yv, zv)
        self.itemData = itemData
        self.count = count

        self.width = itemEntitySize
        self.height = itemEntitySize
        self.rect = pygame.rect.Rect(0, 0, self.width, self.height)

        self.maxFallingVelocity = -10

        self.timers = {
            "pickupDelay": 20
        }



    def positionUpdates(self, player):

        a = findBlock(self.x, self.y - self.height, self.z, ignoreWater = True)
        b = findBlock(self.x, self.y - self.height, self.z + self.width, ignoreWater = True)
        c = findBlock(self.x + self.width, self.y - self.height, self.z, ignoreWater = True)
        d = findBlock(self.x + self.width, self.y - self.height, self.z + self.width, ignoreWater = True)
        blockBelow = False
        if a or b or c or d:
            blockBelow = True
        
        if blockBelow:
            # snap to floor
            self.yv = 0
            # normalize y coordinate, should set it to on top of the block
            self.y = math.floor(self.y / blockSize) * blockSize + self.height

            # friction
            self.xv -= self.xv / 15
            self.zv -= self.zv / 15
            if self.xv > -0.1 and self.xv < 0.1:
                self.xv = 0
            if self.zv > -0.1 and self.zv < 0.1:
                self.zv = 0
        else:
            # do gravity
            if self.yv > self.maxFallingVelocity:
                self.yv -= gravity

        # do wall collision, but only if xv/zv is != 0, for performance
        if self.xv != 0:
            if self.xv > 0:
                blockToRight = findBlock(self.x + self.width + 1, self.y, self.z)
                if blockToRight:
                    self.x -= self.xv
                    self.xv = 0
            if self.xv < 0:
                blockToLeft = findBlock(self.x - 1, self.y, self.z)
                if blockToLeft:
                    self.x -= self.xv
                    self.xv = 0

        if self.zv != 0:
            if self.zv > 0:
                blockToUp = findBlock(self.x, self.y, self.z - 1)
                if blockToUp:
                    self.z -= self.zv
                    self.zv = 0
            if self.zv < 0:
                blockToDown = findBlock(self.x, self.y, self.z + self.width + 1)
                if blockToDown:
                    self.z -= self.zv
                    self.zv = 0

        insideABlock = findBlock(self.x + self.width/2, self.y - self.height/2, self.z + self.width/2, ignoreWater = True)
        if insideABlock:
            self.y += blockSize/2
            self.yv = 2
        
        self.x += self.xv
        self.y += self.yv
        self.z += self.zv

        self.rect.x = self.x
        self.rect.y = self.z

    def runTimers(self):
        for key, value in self.timers.items():
            if value > 0:
                value -= 1
            if value < 0:
                value += 1
            self.timers[key] = value

    def playerInteraction(self, player):
        if self.timers["pickupDelay"] == 0:
            
            if self.rect.colliderect(player.rect):

                itemPickedUp = player.giveItem(self.itemData, self.count)
                if itemPickedUp:
                    self.deleteSelf = True

    def doStuff(self, player):
        
        if not self.deleteSelf:
            self.positionUpdates(player)
            self.runTimers()

            self.playerInteraction(player)








print("entities initialized")

File: Entities.mjs
import { blockSize, gravity, maxStackSize, itemEntitySize, Rect, consoleLog, showLoadingProgress, camera } from "./GlobalVariables.mjs";
import { findBlock } from "./Worldgen.mjs";

showLoadingProgress("loading Entities.mjs");

// base entity with no ai
class Entity {
    constructor(x, y, z, xv, yv, zv) {
        this.x = x || 0; this.xv = xv || 0;
        this.y = y || 0; this.yv = yv || 0;
        this.z = z || 0; this.zv = zv || 0;

        this.deleteMe = false;

        this.width = blockSize;
        this.height = this.width;

        this.rect = Rect(0, 0, this.width, this.height);


    };
};

export class ItemEntity extends Entity {
    constructor(x, y, z, xv, yv, zv, count, itemData) {


        super(x, y, z, xv, yv, zv);

        this.itemData = itemData
        this.count = count

        this.width = itemEntitySize;
        this.height = itemEntitySize;
        this.rect = Rect(0, 0, this.width, this.height);

        this.renderData = {
            "drawType": "image",
            "position": [0, 0],
            "width": this.width,
            "height": this.height
        }

        this.terminalVelocity = -10;

        this.timers = {
            "pickupDelay": 20
        }




        this.positionUpdates = function (player) {

            let a = findBlock(this.x, this.y - this.height, this.z, ignoreWater = true)
            let b = findBlock(this.x, this.y - this.height, this.z + this.width, ignoreWater = true)
            let c = findBlock(this.x + this.width, this.y - this.height, this.z, ignoreWater = true)
            let d = findBlock(this.x + this.width, this.y - this.height, this.z + this.width, ignoreWater = true)
            let blockBelow = false;

            if (a || b || c || d) { blockBelow = true; };

            if (blockBelow) {
                this.yv = 0;

                // snap this onto the block it should be on
                this.y = Math.floor(this.y / blockSize) * blockSize + this.height;

                // do friction
                this.xv -= this.xv / 15;
                this.zv -= this.zv / 15;

                if (Math.abs(this.xv) < 0.1) { this.xv = 0; };
                if (Math.abs(this.zv) < 0.1) { this.zv = 0; };

            } else {
                if (this.yv > this.terminalVelocity) { this.yv -= gravity; };
            };


            if (this.xv != 0) {

                let sideValue = this.x;
                let blockToSide = false;

                if (this.xv > 0) { sideValue += this.width; };

                let collision = {
                    "corner 1": findBlock(sideValue, this.y, this.z),
                    "corner 2": findBlock(sideValue, this.y, this.z + this.width),
                    "corner 3": findBlock(sideValue, this.y - this.height, this.z),
                    "corner 4": findBlock(sideValue, this.y - this.height, this.z + this.width),
                };

                for (const cornerPassedCollision of Object.values(collision)) {
                    if (cornerPassedCollision) { blockToSide = true; break; }
                };


                if (blockToSide) { this.x -= this.xv; this.xv = 0; };
            };


            if (this.zv != 0) {

                let sideValue = this.z;
                let blockToSide = false;

                if (this.zv > 0) { sideValue += this.width; };

                let collision = {
                    "corner 1": findBlock(this.x, this.y, sideValue),
                    "corner 2": findBlock(this.x + this.width, this.y, sideValue),
                    "corner 3": findBlock(this.x, this.y - this.height, sideValue),
                    "corner 4": findBlock(this.x + this.width, this.y - this.height, sideValue),
                };

                for (const cornerPassedCollision of Object.values(collision)) {
                    if (cornerPassedCollision) { blockToSide = true; break; }
                }


                if (blockToSide) { this.z -= this.zv; this.zv = 0; };
            };
        };

        let insideABlock = findBlock(this.x + this.width / 2, this.y - this.height / 2, this.z + this.width / 2, ignoreWater = true)
        if (insideABlock) { this.y += blockSize / 2; };


        // update image things
        
        let y = 0;
        y = Math.floor(this.y / blockSize);
        if (y >= chunkSize[1]) {y = chunkSize[1] - 1};
        if (y < 0) {y = 0};

        this.renderData.x = this.x - camera.x;
        this.renderData.y = this.z - camera.z;
        this.renderData.yLayer = y;

        
        this.x += this.xv;
        this.y += this.yv;
        this.z += this.zv;

        this.rect.x = this.x;
        this.rect.y = this.z;


        this.runTimers = function () {
            for (const key of Object.keys(this.timers)) {
                let value = this.timers[key];

                if (value > 0) { value -= 1; };
                if (value < 0) { value += 1; };

                this.timers[key] = value;
            };
        };

        this.playerInteraction = function (player) {
            if (this.timers.pickupDelay == 0) {

                if (this.rect.collide.rect(player.rect)) {

                    let itemPickedUp = player.giveItem(this.itemData, this.count);
                    if (itemPickedUp) { this.deleteMe = true; }
                };
            };
        };

        this.doStuff = function (player) {

            if (!this.deleteMe) {
                this.positionUpdates(player);
                this.runTimers();
                this.playerInteraction(player);
            };
        };
    };
};









showLoadingProgress("Entities.mjs initialized");
